lib/src/css_mirroring/bidi_css_generator.dart

```diff
@@ -0,0 +1,136 @@
+// Copyright 2015 Google Inc. All Rights Reserved.
+
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+library scissors.src.css_mirroring.bidi_css_generator;
+
+import 'dart:async';
+
+import 'package:csslib/parser.dart' show parse;
+import 'package:csslib/visitor.dart'
+    show
+    RuleSet,
+    StyleSheet,
+    TreeNode,
+    Declaration,
+    Directive,
+    MediaDirective,
+    HostDirective,
+    PageDirective,
+    CharsetDirective,
+    FontFaceDirective,
+    ImportDirective,
+    NamespaceDirective;
+import 'package:quiver/check.dart';
+import 'package:source_maps/refactor.dart';
+import 'package:source_span/source_span.dart';
+
+import '../utils/enum_parser.dart';
+import 'util_classes.dart';
+import 'transformer.dart' show CssMirroringSettings, Direction, flipDirection;
+
+part 'directive_processors.dart';
+
+part 'rulesets_processor.dart';
+
+part 'util_functions.dart';
+
+/// Type of a function that does LTR/RTL mirroring of a css.
+typedef Future<String> CssFlipper(String inputCss);
+
+/// BidiCssGenerator generates a CSS which comprises of orientation neutral,
+/// orientation specific and flipped orientation specific parts.
+///
+/// See BidirectionalCss.md for more details.
+
+class BidiCssGenerator {
+  final String _originalCss;
+  final String _flippedCss;
+  final String _cssSourceId;
+  final Direction _nativeDirection;
+
+  FlippableEntities<TreeNode> _topLevels;
+
+  BidiCssGenerator._(this._originalCss, this._flippedCss, this._cssSourceId,
+      this._nativeDirection) {
+    _topLevels = new FlippableEntities(
+        parse(_originalCss).topLevels,
+        parse(_flippedCss).topLevels
+    );
+  }
+
+  static build(String originalCss, String cssSourceId,
+      Direction nativeDirection, CssFlipper cssFlipper) async {
+    return new BidiCssGenerator._(originalCss, await cssFlipper(originalCss),
+        cssSourceId, nativeDirection);
+  }
+
+
+  /// Main function which returns the bidirectional CSS.
+  String getOutputCss() {
+    var orientationNeutralTransaction = _makeTransaction(
+        _originalCss, _cssSourceId);
```

Comment:
nit: when breaking lines, prefer to break higher in the ast:

    var orientationNeutralTransaction =
        _makeTransaction(_originalCss, _cssSourceId);